middle_end/flambda/lift_code.ml

```diff
@@ -17,102 +17,339 @@
+        Misc.fatal_errorf "@[<hv>Attempting to let-bind@ %a@]"
+          Flambda.print inner_expr;
+    end;
+    defs, Var var
+  | (Mutable _ | Region) :: _ | [] ->
+    Misc.fatal_errorf "Expected binding for %a"
+      Variable.print var
+
+let rebuild_expr defs var =
+  let defs, body = split_defs defs var in
+  rebuild_let defs body
+
+let region_delta defs =
+  List.fold_left
+    (fun delta def ->
+       match def with
+       | Exclave -> delta - 1
+       | Region -> delta + 1
+       | Immutable _ | Mutable _ -> delta)
+    0 defs
+
+let defs_open_region defs =
+  region_delta defs > 0
+
+let defs_close_region defs =
+  region_delta defs < 0
+
+let check_defs defs =
+  if check_invariants () then
+    assert (not (defs_open_region defs))
+
+type yes_no_or_maybe = Yes | No | Maybe
+
+let should_lift_regions ~in_closure =
+  match in_closure with
+  | No | Maybe ->
+      (* Don't lift regions if we may be outside of any closure. Doing so would
+         create top-level exclaves, which interact badly with
+         [Lift_let_to_initialize_symbol]. *)
+      false
+  | Yes -> true
+
+let rec extract_let_expr (acc:def list) dest let_expr ~in_closure =
   let module W = Flambda.With_free_variables in
   let acc =
-    match let_expr with
+    match (let_expr : Flambda.let_expr) with
     | { var = v1; defining_expr = Expr (Let let2); _ } ->
-        let acc, body2 = extract_let_expr acc let2 in
-        Immutable(v1, W.expr body2) :: acc
+        extract_let_expr acc v1 let2 ~in_closure
     | { var = v1; defining_expr = Expr (Let_mutable let_mut); _ } ->
-        let acc, body2 = extract_let_mutable acc let_mut in
-        Immutable(v1, W.expr body2) :: acc
+        extract_let_mutable acc v1 let_mut ~in_closure
+    | { var = v1; defining_expr = Expr (Region expr); _ }
+          when should_lift_regions ~in_closure ->
+        extract_region acc v1 expr ~in_closure
+    | { var = v1;
+        defining_expr = Expr (Apply ({ reg_close = Rc_close_at_apply; _ }
+                                     as apply)) }
+          when should_lift_regions ~in_closure ->
+        extract_tail_call acc v1 apply
+    | { var = v1;
+        defining_expr = Expr (Send ({ reg_close = Rc_close_at_apply; _ }
+                                    as send)) }
+          when should_lift_regions ~in_closure ->
+        extract_tail_send acc v1 send
     | { var = v; _ } ->
         Immutable(v, W.of_defining_expr_of_let let_expr) :: acc
   in
   let body = W.of_body_of_let let_expr in
-  extract acc body
+  extract acc dest body ~in_closure
 
-and extract_let_mutable acc (let_mut : Flambda.let_mutable) =
+and extract_let_mutable acc dest let_mut ~in_closure =
   let module W = Flambda.With_free_variables in
   let { Flambda.var; initial_value; contents_kind; body } = let_mut in
   let acc = Mutable(var, initial_value, contents_kind) :: acc in
-  extract acc (W.of_expr body)
+  extract acc dest (W.of_expr body) ~in_closure
+
+(* C[let x = region M in x] *)
```

Comment:
I don't think the body of the let is constrained to be `x` here?

Reply:
Ah, this isn't very clear: `body` is the body of the region, not the body of the `let`. So it's `M` in the expression.

Would these comments be any clearer if they read
`<acc>[let <dest> = region <body> in <dest>]` and such? I find it helpful to be able to map a notional equational theory onto the implementation code, but clearly I went too far and tried to write a POPL paper in the comments ...